### ISSUE TITLE
Reduce Packagist footprint

### ISSUE THREAD

**Issue by scriptPilot** [mevdschee/php-crud-api#948]
I propose to reduce the files published in Packagist and installed via Composer:

<img width="259" alt="Bildschirmfoto 2023-01-08 um 00 34 28" src="https://user-images.githubusercontent.com/19615586/211174076-77bb4eb6-ddfc-4abd-a83e-b9c763fecf30.png">

**Comment by mevdschee** [comment]
>I propose to reduce the files published in Packagist and installed via Composer:

For development purposes most of these files are useful and that's also what composer is used for. I know there is a "--dev" flag in composer. Is there a way we can use that?

**Comment by scriptPilot** [comment]
In my understanding, repositories should contain only these files, which are absolutely relevant to use the class. If anybody wants to develop on a class, the repository should be cloned. Otherwise the vendor folder will blow up very quickly 😬

**Comment by scriptPilot** [comment]
PS: All other classes I use with Composer contain only the consolidated class files.

**Comment by scriptPilot** [comment]
Regarding the dev flag, it is more about dependency management I understand:
https://stackoverflow.com/questions/19117871/what-is-the-difference-between-require-and-require-dev-sections-in-composer-json

What I am asking for is to reduce the package footprint 🤓

**Comment by scriptPilot** [comment]
In your repository, only the `src` might be relevant ...

**Comment by mevdschee** [comment]
>In your repository, only the src folder might be relevant ...

Can you propose an improvement to the `composer.json` that limits the composer package to this folder (and the license and readme)?

**Comment by scriptPilot** [comment]
This thread might be helpful to you:
https://stackoverflow.com/questions/17049313/how-to-ignore-directories-with-composer

**Comment by scriptPilot** [comment]
Composer File:
https://getcomposer.org/doc/04-schema.md#archive

**Comment by mevdschee** [comment]
Can you give a Github link to a composer package that uses this? So that I can inspect and copy their usage of the "archive" feature as it is not entirely clear to me how I should use it.

**Comment by scriptPilot** [comment]
I understand you define the archive.exclude in the composer.json file and Packager will exclude these files when pulling the package updates. Maybe you could give it a try :-)

````json
{
    "name": "mynamespace/mypackage",
    "type": "library",
    "license": "MIT",
    "description": "An example package with archive excludes.",
    "archive": {
        "exclude": [
            ".gitlab-ci.yml",
            ".idea",
            "package.json",
            "package-lock.json"
        ]
    }
}

````

**Comment by mevdschee** [comment]
I checked some repositories from Symfony and also some from Spatie (I'm very fond of their PHP open source contributions) and couldn't find the usage of the "archive" composer feature to exclude tests. Do you have any examples? Are you sure this is a best practice?

**Comment by scriptPilot** [comment]
Mhm ... it seems the documentation is misleading here ([thread](https://stackoverflow.com/questions/20316833/excluding-files-from-composer-packages)). I will try to find an example.

**Comment by scriptPilot** [comment]
It seems the solution with `.gitattributes` is the solution:

Repo:
https://github.com/php-curl-class/php-curl-class

Excludes:
https://github.com/php-curl-class/php-curl-class/blob/master/.gitattributes

**Comment by scriptPilot** [comment]
In my understanding:
- .gitattributes defines which files and folder to exclude
- GitHub is building the ZIP archive
- Packagist is pulling the ZIP archive

<img width="240" alt="Bildschirmfoto 2023-01-11 um 20 13 07" src="https://user-images.githubusercontent.com/19615586/211896928-f71fab88-ab81-428d-9f3d-a7b653afdca0.png">

**Comment by scriptPilot** [comment]
Only the relevant files are zipped and exported 🥳

**Comment by mevdschee** [comment]
I've got these assembled for `.gitattributes` now:

```
/docker/ export-ignore
/examples/ export-ignore
/extras/ export-ignore
/tests/ export-ignore
/.gitattributes export-ignore
/.gitignore export-ignore
/.htaccess export-ignore
/*.php export-ignore
/*.sh export-ignore
/docker-compose.yml export-ignore
/Dockerfile export-ignore
```

I might break some people's update flow as I'm excluding all PHP files in the root (including the single file build), but maybe those should be released as separate files anyway.

What do you think?

**Comment by mevdschee** [comment]
I've noticed that with `git archive v2.14.14 --worktree-attributes  > v2.14.14.tar` I get a reduced tar package from the repo. It reduces the size of the package from 2 MiB to 490 KiB. I'm not sure that that is what packagist is doing, but if it does, then that helps.

**Comment by scriptPilot** [comment]
Well, if somebody is using Composer, probably the PHP files from the root are not relevant 🤔

**Comment by scriptPilot** [comment]
> I've noticed that with `git archive v2.14.14 --worktree-attributes > v2.14.14.tar` I get a reduced tar package from the repo. It reduces the size of the package from 2 MiB to 490 KiB. I'm not sure that that is what packagist is doing, but if it does, then that helps.

I read that Packagist is downloading the ZIP from GitHub.
So it is reduced now to 25% - good job 😜

**Comment by mevdschee** [comment]
> So it is reduced now by 75% - good job stuck_out_tongue_winking_eye

It's not live yet..

**Comment by mevdschee** [comment]
I've pushed the change (added the .gitattributes), but it also affects all people that want to manually download the repository, see:

![image](https://user-images.githubusercontent.com/1288217/211905873-d6212f88-2d5b-4198-92a2-e8f5c9cd4972.png)

**Comment by scriptPilot** [comment]
Mhm, I am thinking about the use cases ....

- quick and dirty: copy the **api.php** file and use it
- using Composer > only the `src` folder should be relevant
- developer > clone the repository to change your code

Maybe I oversee something but normally it should not be an issue for on of these use cases.

**Comment by scriptPilot** [comment]
In your installation instruction you explain no. 1 and no. 2 - I think the audience for no. 3 should know about cloning 🤓

**Comment by scriptPilot** [comment]
Finally, the problem is that there is a huge footprint in the **vendor** folder compared to the actual API source code. It is downloaded to the development environment and uploaded to the client server during deployment. For sure, I can modify the deployment script to remove unrelated files and folder before but ideally Composer does install only relevant files 🙃

**Comment by mevdschee** [comment]
>quick and dirty: copy the api.php file and use it

As I see it that means that you need to click the file in the user interface, click the "raw" button, click "save this page" in your browser and store it somewhere?

**Comment by scriptPilot** [comment]
Mmh, you could explain it and simplify by providing the link:

https://raw.githubusercontent.com/mevdschee/php-crud-api/main/api.php

**Comment by cweagans** [comment]
IMO, api.php is a build artifact, not a source file. It would be relatively straightforward to put together a quick Github Actions workflow to build this file and then attach it to the release (whenever one is created).

**Comment by mevdschee** [comment]
>Mmh, you could explain it and simplify by providing the link:

I like the idea of having a download link in the readme.

>IMO, api.php is a build artifact .. put together a quick Github Actions workflow to build this file and then attach it

I like this idea as well and I was looking for examples in other (PHP) repositories. Did you find a good one?

**Comment by cweagans** [comment]
If you can point me toward whatever you use to build api.php, I'll build the workflow for you.

**Comment by mevdschee** [comment]
I've read that artifacts are only stored for 90 days, it that true? It makes me doubt using artifacts.

**Comment by cweagans** [comment]
Artifacts for a workflow run are only retained for a limited time. *Release* artifacts are stored indefinitely.

**Comment by mevdschee** [comment]
>If you can point me toward whatever you use to build api.php, I'll build the workflow for you.

Well, it is mainly `php build.php` as described by the readme (that produces the artifacts).

You may need to run `php install.php` first to install and patch the dependencies.

**Comment by scriptPilot** [comment]
Any update here? After the latest update by Composer I am currently uploading 25 MB of new data to each project 🤪

**Comment by mevdschee** [comment]
I have found that the Github CLI  package provides the "gh" command that allows to write a script that uploads artifacts after build.

**Comment by mevdschee** [comment]
Released in v2.14.19

**Comment by scriptPilot** [comment]
Great improvement - thanks a lot 🥳
